### PR TITLE
enhancement(datadog-grok): extend `keyvalue` filter to match Datadog implementation

### PR DESCRIPTION
**Pull request by vladimir-dd**
Extends `keyvalue` filter to match Datadog [implementation](https://docs.datadoghq.com/service_management/events/pipelines_and_processors/grok_parser/?tab=filters#key-value-or-logfmt).

### CODE REVIEW

File: src/datadog/grok/filters/keyvalue.rs
Comment (bruceg) [inline src/datadog/grok/filters/keyvalue.rs:190]: Personally, I'd do this in the inverse just to avoid the early return, but TIOLI:
```suggestion
                    if !key.contains(' ') {
```

File: src/datadog/grok/filters/keyvalue.rs
Comment (bruceg) [inline src/datadog/grok/filters/keyvalue.rs:214]: I think this inner block is complex enough to warrant its own function. Is that reasonable? i.e. something like `.for_each(|c| filter.try_parse_value(&mut result, c))` or whatever other parameters it needs.

File: src/datadog/grok/filters/keyvalue.rs
Comment (bruceg) [inline src/datadog/grok/filters/keyvalue.rs:306]: Would this make sense to put above in the single branch that creates the `Value::Integer` variant? Also, does this secondary parsing now make this PR a breaking change, as the string `015` would parse as 15 before and 13 after the change?

File: src/datadog/grok/parse_grok.rs
Comment (vladimir-dd) [reply src/datadog/grok/parse_grok.rs:258]: Sure, might be useful.

File: src/datadog/grok/filters/keyvalue.rs
Comment (vladimir-dd) [reply src/datadog/grok/filters/keyvalue.rs:188]: I prefer an early exit to be honest than having an extra nested level, however given that we can extract the code below into a function(as you suggest) this make sense.

File: src/datadog/grok/filters/keyvalue.rs
Comment (vladimir-dd) [reply src/datadog/grok/filters/keyvalue.rs]: Good idea, updated https://github.com/vectordotdev/vrl/pull/1015/commits/401c44eca171b03f1c5ef47379cb622b46f0129e

File: src/datadog/grok/filters/keyvalue.rs
Comment (vladimir-dd) [reply src/datadog/grok/filters/keyvalue.rs]: Good idea - https://github.com/vectordotdev/vrl/pull/1015/commits/41ea1c9b6834d8faeb5481e37332948d149fbd33

File: src/datadog/grok/filters/keyvalue.rs
Comment (vladimir-dd) [reply src/datadog/grok/filters/keyvalue.rs:84]: >Also, [..1] and [1..] will only work if the first character is ASCII
Good catch, yeah, I remember running into this issue once.

Comment (vladimir-dd) [reply src/datadog/grok/filters/keyvalue.rs]: Updated https://github.com/vectordotdev/vrl/pull/1015/commits/44d0801f7c39d5f4934771ea883728d4eca75041

File: src/datadog/grok/filters/keyvalue.rs
Comment (vladimir-dd) [reply src/datadog/grok/filters/keyvalue.rs]: Good idea - https://github.com/vectordotdev/vrl/pull/1015/commits/48370561d9d70e9449ca286780a64f9fc01e95c6

File: src/datadog/grok/filters/keyvalue.rs
Comment (vladimir-dd) [reply src/datadog/grok/filters/keyvalue.rs]: Good idea - https://github.com/vectordotdev/vrl/pull/1015/commits/d6edc7c6aa882b4b96f968249fb8fae0d7701b19

File: src/datadog/grok/filters/keyvalue.rs
Comment (vladimir-dd) [reply src/datadog/grok/filters/keyvalue.rs]: Agree - extracted a function https://github.com/vectordotdev/vrl/pull/1015/commits/b4d67ad379fbb035d8c39c137bc8c894ab7fb1c8.
As to parsing octal numbers - we keep parsing them as decimal(no octal to decimal conversion), which corresponds to Datadog grok implementation.

File: src/datadog/grok/filters/keyvalue.rs
Comment (bruceg) [inline src/datadog/grok/filters/keyvalue.rs:100]: Final wondering, and this is in bike shed territory for sure: Maybe first assign the individual characters, and lift the repeated `.to_string()` calls out of the condition, or even leave them as `char`s when they are passed on.

File: src/datadog/grok/filters/keyvalue.rs
Comment (bruceg) [inline src/datadog/grok/filters/keyvalue.rs:120]: nit: I'm curious why the underscore here? It is used below…

File: src/datadog/grok/filters/keyvalue.rs
Comment (bruceg) [inline src/datadog/grok/filters/keyvalue.rs:200]: Could this be a method on `KeyValueFilter`?

File: src/datadog/grok/filters/keyvalue.rs
Comment (bruceg) [inline src/datadog/grok/filters/keyvalue.rs:227]: The cloning here turns into some nasty O(N^2) behavior for array values, not to mention reinserting the same value at an existing slot. Could you use the `entry` interface to avoid needing to clone and reinsert?

File: src/datadog/grok/filters/keyvalue.rs
Comment (bruceg) [inline src/datadog/grok/filters/keyvalue.rs:297]: I was thinking to move the `validate_octal_number` bits right here so you don't have to re-match on the result value being an `Integer` since you already know it is. I see that doesn't work with the `map` function, is there another `nom` function that would allow returning an error?

**Review by bruceg — CHANGES_REQUESTED** [review]
One blocking concern around array insertion behavior, otherwise just "this could be cleaner" comments.

File: src/datadog/grok/filters/keyvalue.rs
Comment (tessneau) [inline src/datadog/grok/filters/keyvalue.rs:62]: Not necessarily a blocker for this PR but -- Would it make sense to move this and the other similar sections, like the one retrieving the value for `field_delimiters`, into their own functions since they're pretty complex?

**Review by tessneau — APPROVED** [review]
looks good to me 👍  I'll defer to @bruceg for final review here since I'm not too familiar with this repo

**Comment by bruceg** [comment]
This is ready for re-review. I'm not sure what to do about the errors in the benchmark code. It almost needs to be parsed from a VRL expression, but that's a significant overhaul compared to what's there.

**Comment by pront** [comment]
> This is ready for re-review. I'm not sure what to do about the errors in the benchmark code. It almost needs to be parsed from a VRL expression, but that's a significant overhaul compared to what's there.

If it simplifies things, we can remove the benchmark to unblock this PR and add a ticket to re-introduce it as a followup.

**Review by pront — APPROVED** [review]
Thanks, this was a tricky one!

File: src/datadog/grok/filters/keyvalue.rs
Comment (tessneau) [inline src/datadog/grok/filters/keyvalue.rs:168]: is the motivation behind having two seperate implementations organizational? i.e. first one concerning itself with initialization and this one with processing

**Comment by tessneau** [comment]
Nicely done ! Thanks for closing the loop on this one

**Review by tessneau — APPROVED** [review]

File: src/datadog/grok/filters/keyvalue.rs
Comment (bruceg) [reply src/datadog/grok/filters/keyvalue.rs:168]: Pretty much just minimizing the code movement in the diff to make it (a little) easier to review. I can move it if you think it's worth it.

Comment (tessneau) [reply src/datadog/grok/filters/keyvalue.rs:168]: Ah gotcha 👍 feels fine as is, I just hadn't seen that much yet so I was mainly curious

**Pull request merged**
